[Scrum 23/feat] startGame 구현

KSJ0128 opened the pull request:
## startGame 구현
### 📝 요약
게임을 시작하는 startGame 로직을 구현했습니다.

### 게임 시작 전 예외 처리
- 방 인원이 2명 이상인지
- 방 상태가 WAIT인지
- 사용자가 모두 Ready인지
- 게임 시작을 요청한 유저가 방장인지

### 게임 시작
- 방 상태를 PLAY로 변경
- Redis 게임 설정 정보 생성

### Redis 정보 생성
```
private void setRedis(Room startRoom) {
        final String HASH_KEY = "GAME" + startRoom.getId();

        // redis에 Key-Value 형태로 값 저장
        userRoomRepository.findAllByRoom(startRoom).stream()
                .map(UserRoom::getUser)
                .forEach(user -> redisTemplate.opsForHash().put(HASH_KEY, user.getName(), "0"));

        // TTL 설정
        redisTemplate.expire(HASH_KEY, startRoom.getTime() * (startRoom.getRound() + 1), TimeUnit.SECONDS);
    }
``` 
위 코드를 통해 redis에 GAME + 방ID로 유저, 스코어 값을 저장했습니다.<br>

<img width="1418" alt="Screenshot 2025-01-30 at 2 24 54 PM" src="https://github.com/user-attachments/assets/141b6e7b-9436-4445-aee7-9c2733fabb93" />
웹 소켓 클라이언트 코드를 통해 API 테스트를 진행하여 알맞은 리턴 값이 반환됨을 확인했습니다.<br>

<br>
<img width="267" alt="Screenshot 2025-01-30 at 2 18 20 PM" src="https://github.com/user-attachments/assets/22d73644-c53a-4a02-a327-e65f70536f61" />
<br>위 사진처럼 redis에 접속해 키 값을 통해 Value 값이 제대로 생성됐는지 확인했습니다.
<br>또한, 시간이 지난 후 TTL에 의해 자동으로 삭제됨을 확인했습니다.


### 🛠️ 변경 사항
- [x] 새로운 기능 추가
- [ ] 버그 수정
- [x] 코드 리팩토링
- [x] 로직에 영향을 미치지 않는 수정(주석 수정, 오타 수정 등)
- [ ] 문서 수정
- [ ] 테스트 코드
- [x] 파일 구조 변경

KSJ0128 committed:
refact: roomId 유효성 검증

enterRoom, quitRoom 메서드에 사용되는 RoomIdDTO에 id 관련 유효성 검증을 추가합니다.

KSJ0128 committed:
feat: startGame 방 상태 검증

게임 시작 전 방 관련 사항을 검증하는 로직입니다.
- 방 내 인원이 두 명 이상인지
- 방 상태가 WAIT인지

KSJ0128 committed:
feat: startGame 유저 상태 검증

게임 시작 전 유저 관련 사항을 검증하는 로직입니다.
- 사용자가 모두 준비 상태인지
- 게임 시작을 요청한 유저가 방장인지

KSJ0128 committed:
feat: 게임 설정을 담는 RedisGameDTO 구현

각 방의 게임 설정을 담는 RedisGameDTO를 구현했습니다.

KSJ0128 committed:
chore: RedisConfig 생성

Key-Value 직렬화 설정이 필요해 Config 파일을 생성합니다.

KSJ0128 committed:
feat: startGame 기본 로직 구현

startGame 기본 로직입니다.
canStartGame()을 통해 게임 시작 가능 여부를 판단합니다.

KSJ0128 committed:
feat: 방 상태 변경 및 redis 저장

게임 방 상태를 PLAY로 변경합니다.
또한, redis에 Game 설정 정보를 저장합니다.

KSJ0128 committed:
feat: 게임 턴 정보 DTO 구현

게임 턴 정보 관련 DTO를 구현했습니다.
해당 턴의 주제어, 그림을 그리는 유저, 참여자 정답 현황을 담고 있습니다.

KSJ0128 committed:
refact: RedisConfig @Configuration 추가

어노테이션을 통해 Configuration 설정을 해줬습니다.

KSJ0128 committed:
refact: startGame 객체 삭제

roomId만 전달해도 될 것 같아 객체를 삭제하고 roomIdDTO로 대체합니다.

KSJ0128 committed:
refact: WebSocket Origin 변경

기존의 Origin에 localhost:8080, 35.216.51.107:8080을 추가했습니다.

KSJ0128 committed:
refact: 방 인원 검증 로직 수정

If문의 조건이 잘못돼 이를 2명 이상인 경우만 허용하도록 수정했습니다.

KSJ0128 committed:
refact: RedisGameDTO 직렬화 적용

해당 객체가 직렬화되지 않아 @Getter를 추가했습니다.

KSJ0128 committed:
feat: startGame 로직 구현

startGame 로직을 구현했습니다.
방 생성 후 해당 방에 속한 참여자, 로비의 참여자에게 방 변경 사항을 메시지 발송합니다.

KSJ0128 committed:
feat: /startGame 스웨거 적용

GameControllerDocs를 통해 /startGame 스웨거 설정을 적용했습니다.

jin20fd reviewed (commented):
`redisTemplate.opsForHash` 라는게 있대여. value로 게임의 모든 정보를 저장하는 것보다 hset, hget활용해서 key의 값을 hash형태로 저장하면 더 좋을 듯합니당
왜냐면 게임 정보 중에서도 잘 바뀌지 않는 정보도 있고 매턴마다 바꿔야하는 정보도 있어서, 이런 정보들은 value내의 키로 특정 값을 바꿔주도록 하면 더 좋을 것 같아여

jin20fd commented on src/main/java/mana/doodleking/domain/game/GameController.java:
```diff
@@ -0,0 +1,44 @@
+    public void startGame(@Header("userId") Long userId, @Valid RoomIdDTO roomIdDTO) {
+        try {
+            gameService.startGame(userId, roomIdDTO);
+            messageSender.send("/topic/room/" + roomIdDTO.getRoomId(), "Start GAME");
```
게임 시작할 때 게임 에 참여하는 사용자랑 기본 점수 0점으로 보내주어야할 것 같아여

KSJ0128 replied:
앗하 ~ 알겟지비

KSJ0128 commented:
> `redisTemplate.opsForHash` 라는게 있대여. value로 게임의 모든 정보를 저장하는 것보다 hset, hget활용해서 key의 값을 hash형태로 저장하면 더 좋을 듯합니당 왜냐면 게임 정보 중에서도 잘 바뀌지 않는 정보도 있고 매턴마다 바꿔야하는 정보도 있어서, 이런 정보들은 value내의 키로 특정 값을 바꿔주도록 하면 더 좋을 것 같아여

이 부분은 제가 좀 더 찾아보도록 하겠습니다:) 땡베감 ~

-> 수정 완료했습니다:)

KSJ0128 committed:
refact: 디렉토리 이동

KSJ0128 committed:
refact: startGame 초기 값 설정

게임을 시작할 때 게임룸 내 참여자에게 참여자:스코어 기본값을 전달합니다.

이를 위해 GameStatusDTO, PlayerScoreDTO를 생성해 startGame()이 리턴하게 합니다.

또한, setRedis()에서 .opsForHash를 통해 key-value 형태로 값을 저장합니다.

KSJ0128 committed:
feat: HashKey, HashValue 관련 redis 설정 추가

RedisConfig 파일에 Hash Key-Value 관련 설정을 추가했습니다.

KSJ0128 committed:
refact: HASH_KEY 변경

HASH_KEY를 GAME + 방ID로 수정했습니다.

KSJ0128 committed:
feat: Swagger 적용

아래 네 가지 API에 대해 Swagger를 적용했습니다.
/app/startGame,
/app/updateRoom,
/app/enterRoom,
/app/quitRoom,
